# Practicumopdrachten Week 2.3 (eerste kans)

## Dependencies importeren

Kopieer in het codeblok hieronder van het vorige practicum de import-code voor de dependencies die het vaakst worden gebruikt om data in te lezen. Geef er ook de gebruikelijke aliassen aan.<br>
Zet eventuele warnings uit.

In [1]:
import pandas as pd
import sqlite3
import warnings
warnings.simplefilter('ignore')

C:\Users\tjorn\AppData\Local\Temp\ipykernel_23416\2821951215.py:1: DeprecationWarning: 
Pyarrow will become a required dependency of pandas in the next major release of pandas (pandas 3.0),
(to allow more performant data types, such as the Arrow string type, and better interoperability with other libraries)
but was not found to be installed on your system.
If this would cause problems for you,
please provide us feedback at https://github.com/pandas-dev/pandas/issues/54466
        
  import pandas as pd


Zet het bestand go_sales.sqlite in hetzelfde mapje als dit notebook.

## Databasetabellen inlezen

Kopieer in het codeblok hieronder van het vorige practicum de code om een connectie met het bestand go_sales.sqlite te maken.

In [2]:
dbconnection = sqlite3.connect('go_sales.sqlite')

Lees van de ingelezen go_sales-database te volgende tabellen in met behulp van "SELECT * FROM *tabel*".
- product
- product_type
- product_line
- sales_staff
- sales_branch
- retailer_site
- country
- order_header
- order_details
- returned_item
- return_reason
- SALESTargetData

In [3]:
product = pd.read_sql("SELECT * FROM product", dbconnection)
product_type = pd.read_sql("SELECT * FROM product_type", dbconnection)
product_line = pd.read_sql("SELECT * FROM product_line", dbconnection)
sales_staff = pd.read_sql("SELECT * FROM sales_staff", dbconnection)
sales_branch = pd.read_sql("SELECT * FROM sales_branch", dbconnection)
retailer_site = pd.read_sql("SELECT * FROM retailer_site", dbconnection)
country = pd.read_sql("SELECT * FROM country", dbconnection)
order_header = pd.read_sql("SELECT * FROM order_header", dbconnection)
order_details = pd.read_sql("SELECT * FROM order_details", dbconnection)
returned_item = pd.read_sql("SELECT * FROM returned_item", dbconnection)
return_reason = pd.read_sql("SELECT * FROM return_reason", dbconnection)
sales_target_data = pd.read_sql("SELECT * FROM SALES_TARGETData", dbconnection)

Let op! Voor alle onderstaande opdrachten mag je <u>alleen Python</u> gebruiken, <u>geen SQL!</u>

## Selecties op één tabel zonder functies

Geef een overzicht met daarin de producten en hun productiekosten waarvan de productiekosten lager dan 100 dollar en hoger dan 50 dollar ligt. (2 kolommen, 23 rijen)

In [4]:
product['PRODUCTION_COST'] = product['PRODUCTION_COST'].astype(float)

condition = (product['PRODUCTION_COST'] > 50) & (product['PRODUCTION_COST'] < 100)

product.loc[condition, ['PRODUCT_NAME', 'PRODUCTION_COST']]

PRODUCT_NAME  PRODUCTION_COST
13                Blue Steel Putter            60.00
14            Blue Steel Max Putter            81.80
17              Course Pro Golf Bag            80.00
24                  Hibernator Lite            60.00
25                       Hibernator            86.00
28  Hibernator Self - Inflating Mat            78.55
31              Hibernator Camp Cot            65.33
32     Canyon Mule Climber Backpack            62.50
56          Granite Climbing Helmet            52.86
58            Husky Harness Extreme            55.00
69                   Granite Hammer            57.13
71        TrailChef Deluxe Cook Set            85.11
81                        Polar Ice            73.33
83                     Polar Sports            80.00
84                       Polar Wave            60.00
85                    Polar Extreme            72.50
88                     Edge Extreme            80.00
91                        Seeker 35            79.19
92                        Seeker 50            92.58
93           TrailChef Double Flame            75.00
94                   Seeker Extreme            94.12
97                   Glacier Deluxe            56.25
98                      Glacier GPS            78.55

Geef een overzicht met daarin de producten en hun marge waarvan de marge lager dan 20 % of hoger dan 60 % ligt. (2 kolommen, 7 rijen) 

In [5]:
product['MARGIN'] = product['MARGIN'].astype(float)

condition = (product['MARGIN'] < 0.2) | (product['MARGIN'] > 0.6)

product.loc[condition, ['PRODUCT_NAME', 'MARGIN']]

PRODUCT_NAME  MARGIN
29                 Hibernator Pad    0.17
32   Canyon Mule Climber Backpack    0.17
100             BugShield Natural    0.67
101               BugShield Spray    0.67
102         BugShield Lotion Lite    0.70
103              BugShield Lotion    0.63
105             BugShield Extreme    0.63

Geef een overzicht met daarin de landen waar met francs wordt betaald. Sorteer de uitkomst op land.  (1 kolom, 3 rijen)

In [6]:
condition = country['CURRENCY_NAME'] == 'francs'

country.loc[condition, ['COUNTRY']].sort_values(by='COUNTRY')

COUNTRY
7       Belgium
0        France
19  Switzerland

Geef een overzicht met daarin de verschillende introductiedatums waarop producten met meer dan 50% marge worden geïntroduceerd (1 kolom, 7 rijen) 

In [7]:
condition = product['MARGIN'] > 0.5

product.loc[condition, ['INTRODUCTION_DATE']].drop_duplicates()

INTRODUCTION_DATE
4          10-12-2019
6          18-12-2019
7          27-12-2019
9           13-1-2020
14         15-12-2019
79         26-10-2016
100         15-2-2011

Geef een overzicht met daarin het eerste adres en de stad van verkoopafdelingen waarvan zowel het tweede adres als de regio bekend is (2 kolommen, 28 rijen)

In [8]:
sales_branch.loc[(sales_branch['ADDRESS2'].notnull()) & (sales_branch['REGION'].notnull()), ['ADDRESS1', 'CITY']]

ADDRESS1              CITY
10  Prol. Paseo de la Reforma No. 51  Distrito Federal
11               202-2-3 Hyakunincho             Tokyo
12                     543-225 Asahi        Osaka City
13                  2315 Queen's Ave         Melbourne
15             Avenida Paulista, 333         São Paulo
21                    3 Albany Court        Birmingham
27               Singelgravenplein 4         Amsterdam

Geef een overzicht met daarin de landen waar met dollars (dollars of new dollar) wordt betaald. Sorteer de uitkomst op land. (1 kolom, 4 rijen) 

In [9]:
condition = country['CURRENCY_NAME'].str.contains('dollar', case=False)
country.loc[condition, ['COUNTRY']].sort_values(by='COUNTRY')

COUNTRY
6       Australia
15         Canada
3          Taiwan
14  United States

Geef een overzicht met daarin beide adressen en de stad van vestigingen van klanten waarvan de postcode begint met een ‘D’ (van duitsland). Filter op vestigingen die een tweede adres hebben. (3 kolommen, 2 rijen) 

In [10]:
condition = retailer_site['POSTAL_ZONE'].str.startswith('D') & retailer_site['ADDRESS2'].notnull()

retailer_site.loc[condition, ['ADDRESS1', 'ADDRESS2', 'CITY']]

ADDRESS1       ADDRESS2       CITY
205  Röntgenstraße 90  3. Tür rechts  Frankfurt
351   Grubesallee 141       4. Stock    Hamburg

## Selecties op één tabel met functies

Geef het totaal aantal producten dat is teruggebracht (1 waarde) 

In [11]:
returned_item['RETURN_QUANTITY'] = returned_item['RETURN_QUANTITY'].astype(int)

returned_item['RETURN_QUANTITY'].sum()

14306

Geef het aantal regio’s waarin verkoopafdelingen gevestigd zijn. (1 waarde)

In [12]:
sales_branch['REGION'].drop_duplicates().count()

15

Maak 3 variabelen:
- Een met de laagste
- Een met de hoogste
- Een met de gemiddelde (afgerond op 2 decimalen)

marge van producten (3 kolommen, 1 rij) 

In [13]:
marge = product['MARGIN']
min_marge = marge.min()
max_marge = marge.max()
avg_marge = marge.mean()

# omzetten in een dataframe
pd.DataFrame({'min_marge': [min_marge], 'max_marge': [max_marge], 'avg_marge': [round(avg_marge, 2)]})

min_marge  max_marge  avg_marge
0       0.17        0.7        0.4

Geef het aantal vestigingen van klanten waarvan het 2e adres niet bekend is (1 waarde)

In [14]:
condition = retailer_site['ADDRESS2'].isnull()

retailer_site.loc[condition, 'RETAILER_SITE_CODE'].count()

286

Geef de gemiddelde kostprijs van de verkochte producten waarop korting (unit_sale_price < unit_price) is verleend (1 waarde) 

In [15]:
# strings (varchar) omzetten naar floats
order_details['UNIT_PRICE'] = order_details['UNIT_PRICE'].astype(float)
order_details['UNIT_SALE_PRICE'] = order_details['UNIT_SALE_PRICE'].astype(float)
order_details['UNIT_COST'] = order_details['UNIT_COST'].astype(float)

condition = order_details['UNIT_SALE_PRICE'] < order_details['UNIT_PRICE']

order_details.loc[condition, 'UNIT_COST'].mean()


125.42305809556548

Geef een overzicht met daarin het aantal medewerkers per medewerkersfunctie (2 kolommen, 7 rijen) 

In [16]:
sales_staff['POSITION_EN'].value_counts()

POSITION_EN
Level 2 Sales Representative    39
Level 3 Sales Representative    24
Branch Manager                  19
Level 1 Sales Representative    13
District Manager                 4
Regional Manager                 2
General Manager                  1
Name: count, dtype: int64

Geef een overzicht met daarin per telefoonnummer het aantal medewerkers dat op dat telefoonnummer bereikbaar is. Toon alleen de telefoonnummer waarop meer dan 4 medewerkers bereikbaar zijn. (2 kolommen, 10 rijen) 

In [17]:
phone_counts = sales_staff['WORK_PHONE'].value_counts()
filtered_phone_counts = phone_counts[phone_counts > 4].dropna() #dropna() is nodig om de NaN-waarde te verwijderen

filtered_phone_counts

WORK_PHONE
1 (305) 557-4810       6
1 (617) 268-6754       6
+31 (0)20 692 93 94    5
1 (403) 232-5986       5
+(61) 03 2982  4242    5
+(44) 121 3505267      5
1 (310) 281-5722       5
+33 1 68 94 52 20      5
1 (206) 292-0012       5
1 (416) 493-5595       5
Name: count, dtype: int64

## Selecties op meerdere tabellen zonder functies

Geef een overzicht met daarin het eerste adres en de stad van vestigingen van klanten uit ‘Netherlands’ (2 kolommen, 20 rijen) 

In [18]:
# retailer_site['COUNTRY_CODE'] is een foreign key naar country['COUNTRY_CODE']

merged_df = pd.merge(retailer_site, country, on='COUNTRY_CODE')
condition = merged_df['COUNTRY'] == 'Netherlands'

merged_df.loc[condition, ['ADDRESS1', 'CITY']]


ADDRESS1          CITY
164                   Beets 36         Beets
165               Perestraat 1      Deventer
166  Laan van Meerdervoort 966      Den Haag
167             Kalfjeslaan 47    Amstelveen
168               Startbaan 11    Amstelveen
169      Noorderspoorsingel 28     Groningen
170              Hogehilweg 13     Amsterdam
171             Spoorstraat 43    Varsseveld
173             Westhavenweg 8     Amsterdam
174              Westzeedijk 4        Muiden
175            Gooiergracht 91         Laren
176            Hoofdstraat 150   Pieterburen
177                Olde Diek 2         Laren
178     Amsterdamsestraatweg 8      Maarssen
179        Krayenhoffstraat 11     Amsterdam
180         Overschiestraat 19     Amsterdam
181              Havenbuurt 12        Marken
182               Duivendaal 1    Wageningen
184              Voorstraat 37  Nieuweschans
185         Stadhouderslaan 18     Rotterdam

Geef een overzicht met daarin de productnamen die tot het producttype ‘Eyewear’ behoren. (1 kolom, 5 rijen) 

In [19]:
merged_df = pd.merge(product, product_type, on='PRODUCT_TYPE_CODE')
condition = merged_df['PRODUCT_TYPE_EN'] == 'Eyewear'
merged_df.loc[condition, 'PRODUCT_NAME']

80        Polar Sun
81        Polar Ice
83     Polar Sports
84       Polar Wave
85    Polar Extreme
Name: PRODUCT_NAME, dtype: object

Geef een overzicht met daarin alle unieke eerste adressen van klantvestigingen en de voornaam en achternaam van de verkopers die ‘Branch Manager’ zijn en aan deze vestigingen hebben verkocht (3 kolommen, 1 rij) 

In [20]:
merged_df_retailer_sales = pd.merge(retailer_site, order_header, on='RETAILER_SITE_CODE')

merged_df_sales_staff = pd.merge(merged_df_retailer_sales, sales_staff, on='SALES_STAFF_CODE')

condition = merged_df_sales_staff['POSITION_EN'] == 'Branch Manager'

merged_df_sales_staff.loc[condition, ['ADDRESS1', 'FIRST_NAME', 'LAST_NAME']].drop_duplicates()

ADDRESS1 FIRST_NAME LAST_NAME
3058  Alameda Santos, 9876     Bayard     Lopes

Geef een overzicht met daarin van de verkopers hun functie en indien zij iets hebben verkocht de datum waarop de verkoop heeft plaatsgevonden. Laat alleen de verschillende namen van de posities zien van de verkopers die het woord ‘Manager’ in hun positienaam hebben staan. (2 kolommen, 9 rijen) 

In [21]:
merged_df = pd.merge(order_header, sales_staff, on='SALES_STAFF_CODE', how="outer")

condition = merged_df['POSITION_EN'].str.contains('Manager', case=False)

merged_df.loc[condition,['POSITION_EN', 'ORDER_DATE']].drop_duplicates()

POSITION_EN  ORDER_DATE
421     Branch Manager         NaN
422    General Manager         NaN
1276  Regional Manager         NaN
1643  District Manager         NaN
5325    Branch Manager  2022-09-09
5326    Branch Manager  2022-10-31
5327    Branch Manager  2022-12-13
5328    Branch Manager  2022-03-13
5329    Branch Manager  2022-01-09

Geef een overzicht met daarin de verschillende namen van producten en bijbehorende namen van producttypen van de producten waarvoor ooit meer dan 750 stuks tegelijk verkocht zijn. (2 kolommen, 9 rijen) 

In [22]:
merged_df_product_order = pd.merge(product, order_details, on='PRODUCT_NUMBER')

merged_df = pd.merge(merged_df_product_order, product_type, on='PRODUCT_TYPE_CODE')

condition = merged_df['QUANTITY'] > 750

merged_df.loc[condition, ['PRODUCT_NAME', 'PRODUCT_TYPE_EN']].drop_duplicates()

PRODUCT_NAME    PRODUCT_TYPE_EN
8616                Star Peg              Tents
15975              Firefly 2           Lanterns
37903        BugShield Spray  Insect Repellents
38313  BugShield Lotion Lite  Insect Repellents
38770       BugShield Lotion  Insect Repellents
39618      BugShield Extreme  Insect Repellents
40754         Sun Shelter 15          Sunscreen
41106         Sun Shelter 30          Sunscreen
41681             Sun Shield          Sunscreen

Geef een overzicht met daarin de productnamen waarvan ooit meer dan 40% korting is verleend. De formule voor korting is: (unit_price - unit_sale_price) / unit_price (1 kolom, 8 rijen) 

In [23]:
merged_df = pd.merge(product, order_details, on='PRODUCT_NUMBER')

merged_df['DISCOUNT'] = ((merged_df['UNIT_PRICE'] - merged_df['UNIT_SALE_PRICE']) / merged_df['UNIT_PRICE']) * 100

merged_df.loc[merged_df['DISCOUNT'] > 40, 'PRODUCT_NAME'].drop_duplicates()

2765     Hailstorm Titanium Woods Set
37446               BugShield Natural
37967                 BugShield Spray
38306           BugShield Lotion Lite
38772                BugShield Lotion
39561               BugShield Extreme
40799                  Sun Shelter 15
41194                  Sun Shelter 30
Name: PRODUCT_NAME, dtype: object

Geef een overzicht met daarin de retourreden van producten waarvan ooit meer dan 90% van de aangeschafte hoeveelheid is teruggebracht (return_quantity/quantity). (1 kolom, 3 rijen) 

In [24]:
merged_df = pd.merge(order_details, returned_item, on='ORDER_DETAIL_CODE')

merged_df['RETURN_PERCENTAGE'] = (merged_df['RETURN_QUANTITY'] / merged_df['QUANTITY']) * 100

condition = merged_df['RETURN_PERCENTAGE'] > 90

merged_df.loc[condition, 'RETURN_REASON_CODE'].drop_duplicates()

0     4
1     5
15    3
Name: RETURN_REASON_CODE, dtype: object

## Selecties op meerdere tabellen met functies

Geef een overzicht met daarin per producttype het aantal producten die tot dat producttype behoren. (2 kolommen, 21 rijen) 

In [25]:
merged_df = pd.merge(product, product_type, on='PRODUCT_TYPE_CODE')

merged_df['PRODUCT_TYPE_EN'].value_counts()

PRODUCT_TYPE_EN
Lanterns                12
Cooking Gear            10
Climbing Accessories     7
Sleeping Bags            7
Tents                    6
Packs                    6
Tools                    6
Insect Repellents        5
Knives                   5
Eyewear                  5
Watches                  5
Sunscreen                5
First Aid                5
Safety                   4
Golf Accessories         4
Woods                    4
Binoculars               4
Navigation               4
Irons                    4
Rope                     4
Putters                  3
Name: count, dtype: int64

Geef een overzicht met daarin per land het aantal vestigingen van klanten die zich in dat land bevinden. (2 kolommen, 21 rijen) 

In [26]:
merged_df = pd.merge(retailer_site, country, on='COUNTRY_CODE')

merged_df['COUNTRY'].value_counts()

COUNTRY
United States     101
Germany            32
Canada             31
France             29
Japan              25
United Kingdom     25
Netherlands        20
Switzerland        13
Sweden             12
Italy              12
Korea              10
Belgium            10
China              10
Taiwan             10
Austria            10
Australia          10
Mexico             10
Spain               9
Denmark             6
Finland             5
Brazil              1
Name: count, dtype: int64

Geef een overzicht met daarin van de producten behorend tot het producttype ‘Cooking Gear’ per productnaam de totaal verkochte hoeveelheid en de gemiddelde verkoopprijs. Sorteer de uitkomst op totaal verkochte hoeveelheid. (4 kolommen, 10 rijen) 

In [27]:
merged_df = pd.merge(product, product_type, on='PRODUCT_TYPE_CODE')
merged_df = pd.merge(merged_df, order_details, on='PRODUCT_NUMBER')

condition = merged_df['PRODUCT_TYPE_EN'] == 'Cooking Gear'

temp_df = merged_df.loc[condition, ['PRODUCT_NAME', 'QUANTITY', 'UNIT_SALE_PRICE']].groupby('PRODUCT_NAME').agg({'QUANTITY': 'sum', 'UNIT_SALE_PRICE': 'mean'}).sort_values(by='QUANTITY')

# herbenoemen van kolommen
temp_df.rename(columns={'QUANTITY': 'TOTAL_QUANTITY', 'UNIT_SALE_PRICE': 'AVG_SALE_PRICE'})


TOTAL_QUANTITY  AVG_SALE_PRICE
PRODUCT_NAME                                             
TrailChef Deluxe Cook Set            3744      120.303667
TrailChef Kettle                     9900       12.130783
TrailChef Single Flame              10614       63.572397
TrailChef Cook Set                  12700       51.204554
TrailChef Double Flame              14510      131.348013
TrailChef Utensils                  15928       17.132397
TrailChef Kitchen Kit               25964       22.210198
TrailChef Canteen                   31712       11.930119
TrailChef Cup                       31834        6.945976
TrailChef Water Bag                 41770        5.999266

Geef een overzicht met daarin van de verkopers behorend tot de verkoopafdeling uit ‘Amsterdam’ de voornaam en achternaam van de verkoper en zijn laagste en hoogste verkooptarget van 2006. (4 kolommen, 4 rijen) 

In [29]:
merged_df = pd.merge(sales_staff, sales_branch, on='SALES_BRANCH_CODE')
merged_df = pd.merge(merged_df, sales_target_data, on='SALES_STAFF_CODE')

condition = (merged_df['CITY'] == 'Amsterdam') & (merged_df['SALES_YEAR'] == 2006)

merged_df = merged_df.loc[condition, ['FIRST_NAME', 'LAST_NAME', 'SALES_TARGET', 'SALES_TARGET']].groupby(['FIRST_NAME', 'LAST_NAME'])

merged_df.agg({'SALES_TARGET': ['min', 'max']})



AttributeError: 'DataFrame' object has no attribute 'name'

Geef een overzicht met daarin per land de naam van het land, de naam van de stad waar de verkoopafdeling is gevestigd (noem de kolomnaam in het overzicht ‘verkoper’) en het aantal steden waar zich klanten bevinden in dat land (noem de kolomnaam in het overzicht ‘klanten’) (3 kolommen, 29 rijen) 

In [30]:
# TODO 28 ipv 29 rijen
merged_df = pd.merge(sales_branch, retailer_site, on='COUNTRY_CODE', suffixes=('_sales_branch', '_retailer_site'))

merged_df = pd.merge(merged_df, country, on='COUNTRY_CODE')

merged_df.groupby(['COUNTRY', 'CITY_sales_branch']).agg({'CITY_retailer_site': 'nunique'}).reset_index()





COUNTRY CITY_sales_branch  CITY_retailer_site
0        Australia         Melbourne                   6
1          Austria              Wien                   9
2          Belgium          Heverlee                   8
3           Brazil         São Paulo                   1
4           Canada           Calgary                  16
5           Canada           Toronto                  16
6            China          Shanghai                   7
7          Finland            Kuopio                   5
8           France              Lyon                  14
9           France             Paris                  14
10         Germany           Hamburg                  18
11         Germany           München                  18
12           Italy            Milano                   9
13           Japan        Osaka City                   5
14           Japan             Tokyo                   5
15           Korea             Seoul                   6
16          Mexico  Distrito Federal                   6
17     Netherlands         Amsterdam                  15
18           Spain            Bilbao                   5
19          Sweden             Kista                  10
20     Switzerland            Zürich                   9
21          Taiwan         San Chung                   9
22  United Kingdom        Birmingham                   9
23  United Kingdom            London                   9
24   United States            Boston                  62
25   United States       Los Angeles                  62
26   United States             Miami                  62
27   United States           Seattle                  62

## Pythonvertalingen van SUBSELECT en UNION met o.a. for-loops

Geef een overzicht met daarin de voornaam en de achternaam van de medewerkers die nog nooit wat hebben verkocht (2 kolommen, 25 rijen) 

In [31]:
# alle medewerkers
all_staff = sales_staff['SALES_STAFF_CODE']

# Lijst met medewerkers die iets hebben verkocht
sales_staff_with_sales = order_header['SALES_STAFF_CODE'].unique()

# Lijst met medewerkers die nog nooit iets hebben verkocht
staff_without_sales = [staff for staff in all_staff if staff not in sales_staff_with_sales]

# dataframe met de voornaam en achternaam van de medewerkers die nog nooit iets hebben verkocht
sales_staff.loc[sales_staff['SALES_STAFF_CODE'].isin(staff_without_sales), ['FIRST_NAME', 'LAST_NAME']]


FIRST_NAME      LAST_NAME
7      Walter         Taylor
8    Penelope        Tamrine
11      Polly         Linton
19      Gracy        Gellens
26       Else         Mörike
27      Frank          Fuchs
30      Fritz         Hirsch
31       Jörg          Kunze
33      Maria      Iacobucci
38       Kick        Kalkman
43      Karin      Bergström
44      Denis           Pagé
47      Sally          White
49      Frank        Bretton
50    Georges  Saint-Germain
56       Alex      Rodriguez
60   Samantha          Floyd
65      Julie          Olsen
69      Corey         Wright
76    Estelle        Johnson
77       Greg        Belding
80      Julie        Béchard
84      Lucía          Reyna
89     Kazumi        Uragome
91      Alice         Walter

Geef een overzicht met daarin het aantal producten waarvan de marge lager is dan de gemiddelde marge van alle producten samen. Geef in het overzicht tevens aan wat de gemiddelde marge is van dit aantal producten waarvan de marge lager dan de gemiddelde marge van alle producten samen is. (1 kolom, 2 rijen) 

In [32]:
# alle producten
all_products = product['PRODUCT_NUMBER']

# gemiddelde marge van alle producten
avg_marge_all_products = product['MARGIN'].mean()

# producten met een marge lager dan de gemiddelde marge van alle producten
products_with_lower_marge = product.loc[product['MARGIN'] < avg_marge_all_products, 'PRODUCT_NUMBER']

# gemiddelde marge van producten met een marge lager dan de gemiddelde marge van alle producten
avg_marge_products_with_lower_marge = product.loc[product['PRODUCT_NUMBER'].isin(products_with_lower_marge), 'MARGIN'].mean()

# dataframe met het aantal producten waarvan de marge lager is dan de gemiddelde marge van alle producten samen en de gemiddelde marge van dit aantal producten
pd.DataFrame({'products_with_lower_marge': [products_with_lower_marge.count()], 'avg_marge_products_with_lower_marge': [avg_marge_products_with_lower_marge]})

products_with_lower_marge  avg_marge_products_with_lower_marge
0                         59                             0.292203

Geef een overzicht met daarin de namen van de producten die voor meer dan 500 (verkoopprijs) zijn verkocht maar nooit zijn teruggebracht. (1 kolom, 13 rijen) 

In [39]:
# producten die voor meer dan 500 (verkoopprijs) zijn verkocht
products_with_high_sales = order_details.loc[order_details['UNIT_SALE_PRICE'] > 500, 'PRODUCT_NUMBER']

# merge order details met returned item
# right join zodat we alle producten hebben die zijn teruggebracht
merged_df = pd.merge(order_details, returned_item, on='ORDER_DETAIL_CODE', how='right')

# producten die nooit zijn teruggebracht
products_never_returned = merged_df['PRODUCT_NUMBER'].loc[merged_df['RETURN_QUANTITY'].isnull()].unique()

# producten die voor meer dan 500 (verkoopprijs) zijn verkocht maar nooit zijn teruggebracht
products_with_high_sales_never_returned = [product for product in products_with_high_sales if product not in products_never_returned]

# series met de namen van de producten die voor meer dan 500 (verkoopprijs) zijn verkocht, maar nooit zijn teruggebracht
product_series = product.loc[product['PRODUCT_NUMBER'].isin(products_with_high_sales_never_returned), 'PRODUCT_NAME']

# series omzetten naar dataframe
product_series.to_frame()

PRODUCT_NAME
3               Hailstorm Steel Irons
4            Hailstorm Titanium Irons
5          Lady Hailstorm Steel Irons
6       Lady Hailstorm Titanium Irons
7        Hailstorm Titanium Woods Set
8           Hailstorm Steel Woods Set
9   Lady Hailstorm Titanium Woods Set
10     Lady Hailstorm Steel Woods Set
19                          Star Dome
20                       Star Gazer 2
21                       Star Gazer 3
22                       Star Gazer 6
55                     Husky Rope 200

Geef een overzicht met daarin per (achternaam van) medewerker of hij/zij manager is of niet, door deze informatie toe te voegen als extra 'Ja/Nee'-kolom.<br>
Hint: gebruik een for-loop waarin je o.a. bepaalt of het woord 'Manager' in de functie (position_en) staat. (2 kolommen, 102 rijen).

In [34]:
for index, row in sales_staff.iterrows():
    if 'Manager' in row['POSITION_EN']:
        sales_staff.loc[index, 'MANAGER'] = True
    else:
        sales_staff.loc[index, 'MANAGER'] = False
        
sales_staff[['LAST_NAME', 'MANAGER']]

LAST_NAME MANAGER
0    Savolainen   False
1           Kim   False
2          Choi   False
3           Kao   False
4          Fang   False
..          ...     ...
97     Iglesias   False
98        Lopes    True
99       Barros   False
100   Guimarães   False
101    Lindholm   False

[102 rows x 2 columns]

Met de onderstaande code laat je Python het huidige jaar uitrekenen.

In [35]:
from datetime import date
date.today().year

2024

Met de onderstaande code selecteer je op een bepaald jaartal uit een datum.

In [36]:
from datetime import datetime

date_str = '16-8-2013'
date_format = '%d-%m-%Y'
date_obj = datetime.strptime(date_str, date_format)

date_obj.year

2013

Geef met behulp van bovenstaande hulpcode een overzicht met daarin op basis van het aantal jaar dat iemand in dienst is of een medewerker ‘kort in dienst’ (minder dan 25 jaar in dienst) of een ‘lang in dienst’ (groter gelijk dan 12 jaar in dienst) is. Geef daarbij per medewerker in een aparte kolom zowel ‘kort in dienst’ als ‘lang in dienst’ aan. Gebruik (wederom) een for-loop.<br>
(2 kolommen, 102 rijen) 

In [37]:
# varchar omzetten naar datetime
current_year = date.today().year

for index, row in sales_staff.iterrows():
    date_hired = datetime.strptime(row['DATE_HIRED'], '%d-%m-%Y')
    years_in_service = current_year - date_hired.year

    if years_in_service < 25:
        sales_staff.loc[index, 'SHORT_IN_SERVICE'] = True
        sales_staff.loc[index, 'LONG_IN_SERVICE'] = False
    elif years_in_service >= 12:
        sales_staff.loc[index, 'SHORT_IN_SERVICE'] = False
        sales_staff.loc[index, 'LONG_IN_SERVICE'] = True